crates/java_string/src/cesu8.rs

```diff
@@ -0,0 +1,298 @@
+                    decoded.extend(surrogate_first.into_iter());
+                }
+            };
+        }
+
+        while let Some(first) = iter.next() {
+            let old_offset = index;
+
+            macro_rules! err {
+                ($error_len:expr) => {
+                    return Err(Utf8Error {
+                        valid_up_to: old_offset,
+                        error_len: $error_len,
+                    })
+                };
+            }
+
+            macro_rules! next {
+                () => {{
+                    index += 1;
+                    match iter.next() {
+                        Some(a) => a,
+                        None => err!(None),
+                    }
+                }};
+            }
+
+            macro_rules! next_cont {
+                ($error_len:expr) => {{
+                    let byte = next!();
+                    if (byte) & !CONT_MASK == TAG_CONT {
+                        byte
+                    } else {
+                        err!($error_len)
+                    }
+                }};
+            }
+
+            if first == 0 {
+                // modified UTF-8 should never contain \0 directly.
+                err!(None);
+            } else if first < 128 {
+                flush_first_surrogate_half!();
+                // Pass ASCII through directly.
+                decoded.push(first);
+            } else if first == 0xc0 {
+                flush_first_surrogate_half!();
+                // modified UTF-8 encoding of null character
+                match next!() {
+                    0x80 => decoded.push(0),
+                    _ => err!(Some(1)),
+                }
+            } else {
+                let w = utf8_char_width(first);
+                let second = next_cont!(Some(1));
+                match w {
+                    // Two-byte sequences can be used directly.
+                    2 => {
+                        flush_first_surrogate_half!();
+                        decoded.extend([first, second].into_iter());
+                    }
+                    3 => {
+                        let third = next_cont!(Some(2));
+                        match (first, second) {
+                            // These are valid UTF-8, so pass them through.
+                            (0xe0, 0xa0..=0xbf)
+                            | (0xe1..=0xec, 0x80..=0xbf)
+                            | (0xed, 0x80..=0x9f)
+                            | (0xee..=0xef, 0x80..=0xbf) => {
+                                flush_first_surrogate_half!();
+                                decoded.extend([first, second, third].into_iter())
+                            }
+                            // First half of a surrogate pair
+                            (0xed, 0xa0..=0xaf) => {
+                                flush_first_surrogate_half!();
+                                surrogate_first = Some([first, second, third]);
+                            }
```

Comment:
I thought about this but I couldn't think of a nice way to do it. Either you have to store the values that you peeked in case it's not the second half, and then change `next!` to take that into account, which I don't think would be faster; or you have to handle all the other cases immediately, which gets a bit nasty once you realize that what you're peeking can itself be the first half of a surrogate pair. Overall I thought this approach was nicer.

Reply:
Hmm, how about this

<details>
    <summary> Code </summary>

```rust
    fn from_modified_utf8_slice(slice: &[u8]) -> Result<JavaString, Utf8Error> {
        let mut offset = 0;
        let mut decoded = Vec::with_capacity(slice.len() + 1);

        while let Some(&first) = slice.get(offset) {
            let old_offset = offset;
            offset += 1;

            macro_rules! err {
                ($error_len:expr) => {
                    return Err(Utf8Error {
                        valid_up_to: old_offset,
                        error_len: $error_len,
                    })
                };
            }

            macro_rules! next {
                () => {{
                    if let Some(&b) = slice.get(offset) {
                        offset += 1;
                        b
                    } else {
                        err!(None)
                    }
                }};
            }

            macro_rules! next_cont {
                ($error_len:expr) => {{
                    let byte = next!();
                    if (byte) & !CONT_MASK == TAG_CONT {
                        byte
                    } else {
                        err!($error_len)
                    }
                }};
            }

            if first == 0 {
                // modified UTF-8 should never contain \0 directly.
                err!(Some(1));
            } else if first < 128 {
                // Pass ASCII through directly.
                decoded.push(first);
            } else if first == 0xc0 {
                // modified UTF-8 encoding of null character
                match next!() {
                    0x80 => decoded.push(0),
                    _ => err!(Some(1)),
                }
            } else {
                let w = utf8_char_width(first);
                let second = next_cont!(Some(1));
                match w {
                    // Two-byte sequences can be used directly.
                    2 => {
                        decoded.extend([first, second]);
                    }
                    3 => {
                        let third = next_cont!(Some(2));
                        match (first, second) {
                            // These are valid UTF-8, so pass them through.
                            (0xe0, 0xa0..=0xbf)
                            | (0xe1..=0xec, 0x80..=0xbf)
                            | (0xed, 0x80..=0x9f)
                            | (0xee..=0xef, 0x80..=0xbf)
                            | (0xed, 0xb0..=0xbf) // <-- Second half of a surrogate pair.
                            => decoded.extend([first, second, third]),
                            // First half of a surrogate pair
                            (0xed, 0xa0..=0xaf) => {
                                // Peek ahead and try to pair the first half of surrogate pair with
                                // second.
                                match &slice[offset..] {
                                    [0xed, fifth @ 0xb0..=0xbf, sixth, ..]
                                        if *sixth & !CONT_MASK == TAG_CONT =>
                                    {
                                        let s = dec_surrogates(second, third, *fifth, *sixth);
                                        decoded.extend(s);
                                        offset += 3;
                                    }
                                    _ => {
                                        // No second half, append the first half directly.
                                        decoded.extend([first, second, third]);
                                    }
                                }
                            }
                            _ => err!(Some(1)),
                        }
                    }
                    _ => err!(Some(1)), // modified UTF-8 doesn't allow width 4
                }
            }
        }
```
</details>

- If it's a second half of a surrogate pair, append it directly to the output.
- If it's a first half of a surrogate pair,
  - Peek ahead for the second half and join with it.
  - Otherwise, append the first half directly to the output.

Now there are fewer branches in the common case.

Reply:
I wanted to keep it as an iterator, but whatever.